File: scripts/fetch_islod_vpo_universities.mjs
/**
 * То же, что fetch_islod_vpo_universities.py: выгрузка ВПО с ISLOD → islodVpoUniversities.js
 * Если API недоступен, собирает запасной список из локальных данных репозитория.
 *
 * Запуск из umirhak2026_front:
 *   node scripts/fetch_islod_vpo_universities.mjs
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.join(__dirname, "..");
const OUT_JS = path.join(ROOT, "src", "data", "islodVpoUniversities.js");

const API = "https://islod.obrnadzor.gov.ru/rlic/api/search";

function writeModule(names, sourceComment) {
  const unique = [...new Set(names.filter(Boolean))].sort((a, b) =>
    a.localeCompare(b, "ru"),
  );
  const lines = unique.map((n) => `  ${JSON.stringify(n)}`).join(",\n");
  const js = `/**
 * Организации ВПО для подсказки «Название вуза» (блок «Поиск диплома»).
 * ${sourceComment}
 */
export const ISLOD_VPO_UNIVERSITY_NAMES = [
${lines}
];
`;
  fs.writeFileSync(OUT_JS, js, "utf8");
  console.log(`Записано ${unique.length} наименований → ${OUT_JS}`);
}

async function fetchFromIslod() {
  const all = [];
  let page = 0;
  const pageSize = 100;
  for (;;) {
    const res = await fetch(API, {
      method: "POST",
      headers: {
        "Content-Type": "application/json;charset=UTF-8",
        "User-Agent": "Mozilla/5.0",
      },
      body: JSON.stringify({
        page,
        size: pageSize,
        sort: { sorted: false, unsorted: true, empty: true },
        query: { region: null, orgName: "", orgType: "VPO" },
      }),
    });
    if (!res.ok) {
      throw new Error(`ISLOD HTTP ${res.status}`);
    }
    const data = await res.json();
    const content = data.content ?? [];
    if (!content.length) break;
    for (const item of content) {
      const name = item.fullName;
      if (name) all.push(name);
    }
    console.log(`Страница ${page}, всего: ${all.length}`);
    page += 1;
    await new Promise((r) => setTimeout(r, 500));
  }
  return all;
}

async function loadFallback() {
  const { RUSSIAN_UNIVERSITIES } = await import("../src/data/russianUniversities.js");
  const { VUZ_LIST_NAMES } = await import("../src/data/vuzList.js");
  const wikiPath = path.join(ROOT, "tmp_wiki_vuz.json");
  let wiki = [];
  try {
    wiki = JSON.parse(fs.readFileSync(wikiPath, "utf8"));
  } catch {
    /* optional */
  }
  return [...RUSSIAN_UNIVERSITIES, ...VUZ_LIST_NAMES, ...wiki];
}

try {
  const fromApi = await fetchFromIslod();
  writeModule(
    fromApi,
    "Источник: API https://islod.obrnadzor.gov.ru/rlic/api/search (orgType=VPO). Пересборка: node scripts/fetch_islod_vpo_universities.mjs",
  );
} catch (e) {
  console.warn("ISLOD недоступен, используется локальный запасной список:", e.message);
  const fb = await loadFallback();
  writeModule(
    fb,
    "Запасной список из src/data (russianUniversities, vuzList, tmp_wiki_vuz). Для полного реестра: python scripts/fetch_islod_vpo_universities.py",
  );
}

File: scripts/fetch_islod_vpo_universities.py
"""
Выгрузка полных наименований организаций ВПО из сводного реестра лицензий Рособрнадзора
и запись в src/data/islodVpoUniversities.js для подсказок в кабинете студента.

Требуется: pip install requests

Запуск из каталога umirhak2026_front:
  python scripts/fetch_islod_vpo_universities.py
"""

from __future__ import annotations

import json
import time
from pathlib import Path

import requests

ROOT = Path(__file__).resolve().parents[1]
OUT_JS = ROOT / "src" / "data" / "islodVpoUniversities.js"
OUT_TXT = ROOT / "universities_ru.txt"

url = "https://islod.obrnadzor.gov.ru/rlic/api/search"

headers = {
    "Content-Type": "application/json;charset=UTF-8",
    "User-Agent": "Mozilla/5.0",
}

page = 0
page_size = 100
all_universities: list[str] = []

while True:
    payload = {
        "page": page,
        "size": page_size,
        "sort": {"sorted": False, "unsorted": True, "empty": True},
        "query": {
            "region": None,
            "orgName": "",
            "orgType": "VPO",
        },
    }

    response = requests.post(url, json=payload, headers=headers, timeout=60)

    if response.status_code != 200:
        print("Ошибка запроса:", response.status_code)
        raise SystemExit(1)

    data = response.json()
    content = data.get("content", [])

    if not content:
        break

    for item in content:
        name = item.get("fullName")
        if name:
            all_universities.append(name)

    print(f"Страница {page} обработана, всего: {len(all_universities)}")

    page += 1
    time.sleep(0.5)

all_universities = sorted(set(all_universities))

with open(OUT_TXT, "w", encoding="utf-8") as f:
    for uni in all_universities:
        f.write(uni + "\n")

lines = ",\n  ".join(json.dumps(name, ensure_ascii=False) for name in all_universities)
js = f"""/**
 * Организации высшего профессионального образования (Рособрнадзор, orgType=VPO).
 * Автогенерация: python scripts/fetch_islod_vpo_universities.py
 */
export const ISLOD_VPO_UNIVERSITY_NAMES = [
  {lines}
];
"""
OUT_JS.write_text(js, encoding="utf-8")

print(f"\nГотово! Найдено вузов: {len(all_universities)}")
print(f"JS: {OUT_JS}")
print(f"TXT: {OUT_TXT}")
